### PR TITLE
Feat: generalize ascat, delly, and manta 

### PR DESCRIPTION
**Pull request by hassanfa**
### This PR:

Changed:
- Generalized ascat, delly, and manta result in workflow.
- Generalized workflow to eliminate duplicate entries and code.

Added:
- Individual rules (i.e. ngs filters) for cnv and sv callers. Only Manta will be delivered and added to the list of output files.
- Added "targeted" and "wgs" tags to variant callers to provide another layer of separation.

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

**Comment by hassanfa** [comment]
1. Tumor-normal WGS:
image

**Comment by hassanfa** [comment]
2. Tumor-normal panel:
image

**Comment by hassanfa** [comment]
3. Tumor-only panel:
image

**Comment by hassanfa** [comment]
4. Tumor-only WGS:
image

**Review by khurrammaqbool — APPROVED** [review]
Looks great!

File: BALSAMIC/utils/rule.py [hunk 1]
```diff
@@ -48,13 +48,15 @@ def get_variant_callers(config, mutation_type: str, mutation_class: str,
         mutation_class: A mutation class string, e.g. somatic
         analysis_type: A analysis type string, e.g. paired
         workflow_solution: A workflow type string, e.g. BALSAMIC
+        sequencing_type: A sequencing type, e.g. wgs or panel
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/utils/rule.py]: fixed!

**Comment by hassanfa** [comment]
Enabled Delly for panel seq as well:
image

and

image

**Comment by hassanfa** [comment]
And all workflows:

image

image

image

image

**Review by ashwini06 — COMMENTED** [review]
Well done 🥇 💯  Complexity has become less complex now :sweat_smile:  I just have minor comments

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 2]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
...
+    conda = config["bioinfo_tools"].get("bcftools"),
+  threads:
+    get_threads(cluster_config, 'ngs_filter_cnvkit')
+  message:
+    "Running ngs filters on annotated CNVKit results."
+  shell:
+    """
+source activate {params.conda};
+
+bcftools view --threads {threads} -f .,PASS -o {output.vcf_sv_pass} -O z {input.vcf};
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:52]: this was intentional, as these filters are experimental. I didn't want to lose any filter with `.` in it.

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 3]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
...
+    conda = config["bioinfo_tools"].get("bcftools"),
+  threads:
+    get_threads(cluster_config, 'ngs_filter_delly')
+  message:
+    "Running ngs filters on annotated CNVKit results."
+  shell:
+    """
+source activate {params.conda};
+
+bcftools view --threads {threads} -f .,PASS -o {output.vcf_sv_pass} -O z {input.vcf};
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:77]: same as previous one.

**Comment by ashwini06** [comment]
@hassanfa : How about the delivery of these latest rules `ngs_filter_delly` and `ngs_filter_ascat`? https://github.com/Clinical-Genomics/BALSAMIC/blob/55063d099dc442da7c9dce4c8d07b299beca0aed/BALSAMIC/commands/report/deliver.py#L94

**Comment by hassanfa** [comment]
> @hassanfa : How about the delivery of these latest rules `ngs_filter_delly` and `ngs_filter_ascat`?
> 
> https://github.com/Clinical-Genomics/BALSAMIC/blob/55063d099dc442da7c9dce4c8d07b299beca0aed/BALSAMIC/commands/report/deliver.py#L94

These are not delivered or stored on purpose. As we need to first to find out how to treat them before storing. Since Khurram just needs results for future investigation.

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 4]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
 bcftools view --threads {threads} -f PASS -o {output.vcf_sv_pass} -O z {input.vcf};
 tabix -p vcf -f {output.vcf_sv_pass};
     """
+
+
+rule ngs_filter_cnvkit:
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:32]: fixed.

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 5]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
...
+    benchmark_dir + 'ngs_filter_' + "{var_type}.somatic.{case_name}.cnvkit.vep.tsv"
+  singularity:
+    Path(singularity_image, config["bioinfo_tools"].get("bcftools") + ".sif").as_posix() 
+  params:
+    case_name = "{case_name}",
+    conda = config["bioinfo_tools"].get("bcftools"),
+  threads:
+    get_threads(cluster_config, 'ngs_filter_cnvkit')
+  message:
+    "Running ngs filters on annotated CNVKit results."
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:47]: fixed.

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 6]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
...
+    benchmark_dir + 'ngs_filter_' + "{var_type}.somatic.{case_name}.delly.vep.tsv"
+  singularity:
+    Path(singularity_image, config["bioinfo_tools"].get("bcftools") + ".sif").as_posix() 
+  params:
+    case_name = "{case_name}",
+    conda = config["bioinfo_tools"].get("bcftools"),
+  threads:
+    get_threads(cluster_config, 'ngs_filter_delly')
+  message:
+    "Running ngs filters on annotated CNVKit results."
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:72]: fixed.

File: BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule [hunk 7]
```diff
@@ -27,3 +27,53 @@ source activate {params.conda};
...
+  shell:
+    """
+source activate {params.conda};
+
+bcftools view --threads {threads} -f .,PASS -o {output.vcf_sv_pass} -O z {input.vcf};
+tabix -p vcf -f {output.vcf_sv_pass};
+    """
+
+
+rule ngs_filter_delly:
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/snakemake_rules/annotation/varcaller_sv_filter.rule:57]: fixed.

File: BALSAMIC/workflows/balsamic.smk [hunk 8]
```diff
@@ -359,16 +349,10 @@ else:
...
                                             vcf=get_vcf(config, ["tnscope"], [config["analysis"]["case_id"]]))])
 
-# SV filters output
+# SV and CNV filters output
 analysis_specific_results.extend([expand(vep_dir + "{vcf}.all.filtered.pass.vcf.gz",
-                                        vcf=get_vcf(config, ["manta"], [config["analysis"]["case_id"]]))])
-
-analysis_specific_results.extend([expand(vcf_dir + "{vcf}.vcf.gz",
-                                         vcf=get_vcf(config, ["delly"],[config["analysis"]["case_id"]]))])
+                                        vcf=get_vcf(config, somatic_caller_sv+somatic_caller_cnv, [config["analysis"]["case_id"]]))])
```
[the remark this answers is not in the document]
Comment (hassanfa) [reply BALSAMIC/workflows/balsamic.smk:354]: fixed

**Pull request merged**